Implement synchronous signal handling in umbrella/worker processes

jschmidt-icinga opened the pull request:
This PR is an attempt to fix #9676.

## Problem

As documented in the issue linked above, on the current master, Signal handling is done in asynchronous handlers installed with `sigaction()`. As per the POSIX standard, these handlers aren't allowed to use any functions that aren't listed as `async-signal-safe`, which we currently violate by calling functions such as `Log()` that perform allocations using `operator new()`.

I've had difficulty reproducing the original problem, even when adding multiple big allocations in the event loops. But even if we can't easily reproduce it, doing these things in signal handlers is still incorrect and should be fixed.

## Considerations

There are two potential solutions to this problem. Either remove all non-`async-signal-safe` functions from the handlers and make them as simple as possible, or make the signal handling synchronous instead, as suggested in #9676 by using `sigtimedwait()`.

Currently there is additional effort involved in setting atomic variables (which is necessary because all threads can potentially receive the signal) that get then queried in the processes main event loops.

Considering this additional effort in re-synchronizing the results from signal handling, the synchronous approach seemed simpler and actually less invasive if we want to still keep a way of logging inside of signal handlers.

The asynchronous handlers for SIGABRT will remain as its critical to what it does.

## Implementation

I moved blocking the relevant signals to earlier in `DeamonCommand::Run()` and now we leave them blocked, instead of unblocking them after forking into "umbrella" and worker processes and installing signal handlers instead. Except for SIGHUP, which gets set to be ignore (SIG_IGN) for the worker, same as before.

From there, on the "umbrella"-side, pending signals are polled individually directly where needed, removing the necessity for all of the global state variables.
https://github.com/Icinga/icinga2/blob/42999027f0c49001d37b3d97927112337458132b/lib/cli/daemoncommand.cpp#L472-L477

However on the worker side, this is not easily possible because the state variables (static members of `Application::` in this case) are shared with the WIN32 code-paths and have other input sources as part of the regular control flow. Therefore a `WorkerSignalHandler()`, similar to the old one in  `lib/cli/daemoncommand.cpp` was added to `lib/base/application.cpp` and replaces the `Utility::Sleep()` in `Application::RunEventLoop()` on non-WIN32 systems.
https://github.com/Icinga/icinga2/blob/42999027f0c49001d37b3d97927112337458132b/lib/base/application.cpp#L474-L478

The function `Application::GetPendingSignal()` contains all the platform-y ugliness to support MacOS and OpenBSD, so the rest of the code doesn't have to deal with it. It also supports being called with an infinite timeout and zero timeout.

## Testing

Manual testing has been done to verify that the signals involved in the umbrella/worker setup and termination (SIGUSR1, SIGUSR2, SIGTERM, SIGINT, SIGHUP) are processed and propagated as expected.

In the future (or pending discussion in the context of this PR), adding automated test cases may be a good idea but considering how entangled these classes and functions currently are this would be difficult without refactoring into more self-contained parts first.

Al2Klimov commented on lib/cli/daemoncommand.cpp:
```diff
@@ -274,12 +323,13 @@ int RunWorker(const std::vector<std::string>& configs, bool closeConsoleLog = fa
-		while (!l_AllowedToWork.load()) {
+		// The umbrella process will send this signal when the previous worker has shut down.
+		while (!GetPendingSignal(SIGUSR2)) {
 			Utility::Sleep(0.2);
```
I'm wondering whether you could do this delay...

Al2Klimov replied:
... as the stuff here sleeps anyway forever, until SIGUSR2. 🤷‍♂️

Al2Klimov commented on lib/cli/daemoncommand.cpp:
```diff
@@ -222,6 +222,55 @@ static double GetDebugWorkerDelay()
+
+	siginfo_t info;
+	timespec ts{.tv_sec = 0, .tv_nsec = 0};
+	int ret = sigtimedwait(&sigSet, &info, &ts);
```
... via timeout of this one. Or you could even use sigwaitinfo()...

jschmidt-icinga replied:
I could, but then I would either have to add a timeout argument to `GetPendingSignal()` or accept that in the other situations where it can and should return immediately also incur this timeout.

Also I think explicitly calling `Utility::Sleep()` makes clearer what is happening than having a magic number float parameter to a function that also does something else.

Using `sigtimedwait()` as a polling function that way is explicitly mentioned in the man page.

Al2Klimov replied:
> then I would either have to add a timeout argument to `GetPendingSignal()`

Exactly. E.g a default of -1 for sigwaitinfo() and values >=0 for sigtimedwait().

> explicitly calling `Utility::Sleep()` makes clearer what is happening

Integrating "(a)wait" in the function's name would either.

Al2Klimov reviewed (commented):
* This doesn't compile on my machine🍎.
* Also, I'm afraid this will break OpenBSD https://man.openbsd.org/sigtimedwait which we explicitly ensured to build #9949.

jschmidt-icinga commented:
> * This doesn't compile on my machine🍎.
> 
>     * Also, I'm afraid this will break OpenBSD https://man.openbsd.org/sigtimedwait which we explicitly ensured to build [GHA: also build on Alpine to test LibreSSL which is used on OpenBSD #9949](https://github.com/Icinga/icinga2/pull/9949).

Well, I've not considered OpenBSD just not implementing POSIX...

It seems you're meant to use `sigwait()` in combination with `timer_settime()` on OpenBSD. I will look into it.

Al2Klimov commented:
FreeBSD _should_ work btw: https://man.freebsd.org/cgi/man.cgi?query=sigwaitinfo&sektion=2&format=html

Al2Klimov commented on lib/base/application.cpp:
```diff
@@ -300,6 +300,130 @@ void Application::SetArgV(char **argv)
+			sigaddset(&sigSet, SIGALRM);
+		}
+
+#if defined(__APPLE__) || defined(__OpenBSD__)
+		ret = sigwait(&sigSet, &info.si_signo);
+		VERIFY(ret != EINVAL);
+#else
+		ret = sigwaitinfo(&sigSet, &info);
+		VERIFY(ret != -1 || errno != EINVAL);
+#endif
```
Also, why (to have to write and maintain) two *nix implementations? Is the BSD one not good enough for Linux? Besides less maintenance overhead, one implementation would be tested by Linux, so we won't have to wait until BSD users complain. (Eat your own 🐶food.)

jschmidt-icinga replied:
The `sigwaitinfo()` implementation also gets the proper siginfo_t information against which we check the parent or worker pids.

If we want to do this identically on all systems, we should either talk about a proper IPC, or just skip the pid-checking part entirely.

Al2Klimov replied:
I see!

> talk about a proper IPC

![](https://i.imgflip.com/3j3klm.jpg)

Al2Klimov replied:
Seriously speaking, we already use AF_*NIX sockets for IPC across fork(2), and this use case where we even check PIDs is literally begging for it.

Al2Klimov reviewed (commented):
> This doesn't compile on my machine🍎.

_Now_ it seems to compile. Though, I recommend you

* to #10424
* and to read https://icinga.com/blog/monitor-all-operating-systems-with-one-solution-icinga-2/

in order to properly test this.

Al2Klimov commented on lib/cli/daemoncommand.cpp:
```diff
@@ -777,12 +665,13 @@ int DaemonCommand::Run(const po::variables_map& vm, const std::vector<std::strin
-			if (termSig != -1) {
-				Log(LogNotice, "cli")
-					<< "Got signal " << termSig << ", forwarding to seamless worker (PID " << currentWorker << ")";
+			auto sig = Application::GetPendingSignal({SIGTERM, SIGINT});
+			if (sig) {
+				UnblockSignalAndSetHandler(sig->si_signo, SIG_DFL);
```
I personally would reset both of them while on it.

jschmidt-icinga replied:
I tried my best to replicate the same behavior as before, where we reset the signal the handler got and nothing else. What is the use-case where we would want to reset SIGINT when we only receive SIGTERM? If there is one and it doesn't hurt anything else, I'd be happy to change this.

Al2Klimov commented on lib/cli/daemoncommand.cpp:
```diff
@@ -620,6 +515,13 @@ int DaemonCommand::Run(const po::variables_map& vm, const std::vector<std::strin
 	SetConsoleOutputCP(65001);
 #endif /* _WIN32 */
 
+#ifndef _WIN32
```
#else seems to make sense here.

jschmidt-icinga replied:
Personally I'd prefer preprocessor #else only for defining alternative implementations of similar/equivalent things. In this case they do completely different things but are just in the same place. I'll think about it.

Al2Klimov commented on lib/cli/daemoncommand.cpp:
```diff
@@ -571,19 +469,16 @@ static pid_t StartUnixWorker(const std::vector<std::string>& configs, bool close
-					Log(LogNotice, "cli")
+				auto sig = Application::GetPendingSignal(SIGUSR2);
+				if (sig && (sig->si_pid == 0 || sig->si_pid == pid)) {
+					Log(LogNotice, "Application")
```
Application? But...

Al2Klimov commented on lib/cli/daemoncommand.cpp:
```diff
@@ -777,12 +665,13 @@ int DaemonCommand::Run(const po::variables_map& vm, const std::vector<std::strin
+			auto sig = Application::GetPendingSignal({SIGTERM, SIGINT});
+			if (sig) {
+				UnblockSignalAndSetHandler(sig->si_signo, SIG_DFL);
+				Log(LogCritical, "cli")
```
... it's still cli here.😕

jschmidt-icinga replied:
Yes, no idea when I changed that or why. I had already seen this and reverted it locally after pushing yesterday, but I wanted to wait for the review and other changes instead of pushing again right away.

Al2Klimov commented on lib/cli/daemoncommand.cpp:
```diff
@@ -328,104 +327,40 @@ int RunWorker(const std::vector<std::string>& configs, bool closeConsoleLog = fa
+ * @param handler The new handler after unblocking (SIG_DFL if omitted)
  */
-static void WorkerSignalHandler(int num, siginfo_t *info, void*)
+static void UnblockSignalAndSetHandler(int signal, void (*handler)(int) = SIG_DFL)
```
Do you actually make use of that default?

jschmidt-icinga replied:
Since I currently call the function with SIG_DFL and SIG_IGN, I could make use of the default, as I intended when I implemented the function, but for readability's sake I'd actually now prefer to always use both arguments, so I'll remove the default value instead.

Al2Klimov commented on lib/cli/daemoncommand.cpp:
```diff
@@ -328,104 +327,40 @@ int RunWorker(const std::vector<std::string>& configs, bool closeConsoleLog = fa
-			break;
-		case SIGUSR2:
-			if (!l_CurrentlyStartingUnixWorkerReady.load()
-				&& (info->si_pid == 0 || info->si_pid == l_CurrentlyStartingUnixWorkerPid.load()) ) {
```
Did this PID check vanish completely?

jschmidt-icinga replied:
That is the check on the side of the umbrella process and it's now done here:
https://github.com/Icinga/icinga2/blob/42999027f0c49001d37b3d97927112337458132b/lib/cli/daemoncommand.cpp#L472-L473

Al2Klimov commented on lib/base/application.cpp:
```diff
@@ -300,6 +300,130 @@ void Application::SetArgV(char **argv)
+		ret = -1;
+		for(auto signal : signals){
+			if (sigismember(&pendingSigs, signal)) {
+				Log(LogCritical, "Application")
+					<< "Found pending signal the hard way";
```
```suggestion
```

jschmidt-icinga replied:
Yes, another oversight. I will fix this in the next version.

jschmidt-icinga commented:
I've just pushed a new version where I addressed some of @Al2Klimov's comments and also added some additional clarification and validation for the timeout passed to `GetPendingSignal()`.

Al2Klimov reviewed (commented):
IIRC, SIGUSR2 is used only once in each direction and only between umbrella and worker process. So it should be as simple as an AF_*NIX socketpair(2)/pipe(2) per worker process: instead of kill(2) you'd just write(2) a single byte (and close(2)?). Then we'd only need one *nix implementation AND we'd know where the IPC comes from, EVEN on OpenBSD.

jschmidt-icinga commented:
@Al2Klimov We're also using the PID for the other signals, even SIGTERM on the worker side.

But I get what you mean. I will talk to the others and see what's their opinion about this. If they think it's worth the effort I'll get to work towards a separate PR, since that would be far outside the scope of this one anyway.

jschmidt-icinga commented:
> Sure, it's an option to keep those two implementations for now – IF you test the final PR version on all different OS to be supported.

I didn't necessarily mean that I want this PR to be merged first. In fact, it makes little sense to merge this if we decide on something else. What I meant is that *if* we end up deciding on a different approach, it should be implemented in a new PR (with this one being closed as a consequence).
